### PR TITLE
[GEOT-7346] Provide a snapshot tag for documentation build

### PR DESCRIPTION
**Pull request by jodygarnett**
[GEOT-7346](https://osgeo-org.atlassian.net/browse/GEOT-7346)

The introduction of a snapshot tag allows install instructions to be written for 2.24-SNAPSHOT nightly build clearly, while still providing clear text for the eventual 2.24.0 release.

Aside: I noticed that the project.version is not being passed successfully from maven to ant to sphinx-build. Apparently the sphinx-build ``-D release=${project.version}`` is used as an override of the ``conf.py`` configuration and is not available to determine the correct build server or source forge download links. Passing this value from ant to sphinx-build as an environmental variable works.
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [ ] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: doc/en/README.md [hunk 1]
```diff
@@ -49,13 +52,72 @@ And can be packaged individually:
...
+Additional targets are available:
+
+    ant developer
+    ant docguide
+
+Customize output with current ``project.version`` and most recent ``project.release`` name:
+
+    ant user -Dproject.version=2.23-SNAPSHOT -Dproject.release=2.23.1
+
+## Python and Sphinx Setup
```
Comment (bradh) [inline doc/en/README.md:80]: Possibly this should be earlier in the README.md

File: doc/en/user/source/conf.py [hunk 2]
```diff
@@ -51,14 +51,24 @@
...
+if project_version == None: 
+  release = '2.24-SNAPSHOT'
+else:
+  release =project_version
 
 # Used in build and documentation links
 branch = 'main'
 
-# Users don't need to see the "SNAPSHOT" notation when it's there
+# Users don't need to see the "SNAPSrHOT" notation when it's there
```
Comment (bradh) [inline doc/en/user/source/conf.py:68]:
```suggestion
# Users don't need to see the "SNAPSHOT" notation when it's there
```

File: doc/en/user/source/conf.py [hunk 3]
```diff
@@ -51,14 +51,24 @@
...
+
+# sphinx-build -D release=${project.version} overrides release configuration
+# but only after conf.py has been used...
+
+# check environmental variable to see if ant build.xml passed in project.version
+project_version = os.getenv("project.version")
+if project_version == None: 
+  release = '2.24-SNAPSHOT'
+else:
+  release =project_version
```
Comment (bradh) [inline doc/en/user/source/conf.py:63]:
```suggestion
  release = project_version
```

File: doc/en/user/source/installation/docker.rst [hunk 4]
```diff
@@ -84,3 +154,45 @@ Here is a list of available extensions (taken from the Geoserver download page):
     dxf          importer        netcdf        vectortiles      ysld
     excel        inspire         ogr-wfs       wcs2_0-eo
 
+Testing Geoserver Community modules
+-----------------------------------
+
+Working with a Nightly build is a good way to test community modules and provide feedback to developers working on new functionality.
+
+To work with a community module you must be using the GeoServer |version|.x nightly build:
```
Comment (bradh) [inline doc/en/user/source/installation/docker.rst:162]:
```suggestion
To work with community modules you must be using the GeoServer |version|.x nightly build that matches the community module build:
```

**Review by bradh — APPROVED** [review]
No problems noted. A couple of minor suggestions. 

However I didn't actually test that it works.

**Comment by jodygarnett** [comment]
I updated the doc guide as well to take advantage of requirement.txt and the new ``ant user-site`` target.

**Comment by aaime** [comment]
The backport to `2.22.x` failed:
```
The process '/usr/bin/git' failed with exit code 1
```
<details>
<summary>stderr</summary>

```
error: could not apply ad676c6235... [GEOT-7346] Pass project.version in to conf.py as an environmental variable to determine snapshot tag
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging doc/en/build.xml
[backport-6812-to-2.22.x 5ca0ec15db] [GEOT-7346] sphinx-autobuild python setup from geowebcache instructions
 Author: Jody Garnett <jody.garnett@gmail.com>
 Date: Fri Apr 28 19:07:08 2023 -0700
 4 files changed, 132 insertions(+), 14 deletions(-)
 create mode 100644 doc/en/.gitignore
 create mode 100644 doc/en/requirements.txt
Auto-merging doc/en/build.xml
Auto-merging doc/en/user/source/conf.py
CONFLICT (content): Merge conflict in doc/en/user/source/conf.py

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-2.22.x 2.22.x
# Navigate to the new working tree
cd .worktrees/backport-2.22.x
# Create a new branch
git switch --create backport-6812-to-2.22.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick cccd4a5448ce90e62f0198a5ae980df94aa4d374,ad676c6235d9b5c7b69410c8c6dc402f37af686e,e4139e5814b0425962b13952e0eb96c107fcfa2f,c732de58d35b2149293d18a39faaf14704d4dbda
# Push it to GitHub
git push --set-upstream origin backport-6812-to-2.22.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-2.22.x
```
Then, create a pull request where the `base` branch is `2.22.x` and the `compare`/`head` branch is `backport-6812-to-2.22.x`.

**Comment by aaime** [comment]
The backport to `2.23.x` failed:
```
The process '/usr/bin/git' failed with exit code 1
```
<details>
<summary>stderr</summary>

```
error: could not apply ad676c6235... [GEOT-7346] Pass project.version in to conf.py as an environmental variable to determine snapshot tag
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging doc/en/build.xml
[backport-6812-to-2.23.x ea6a25f947] [GEOT-7346] sphinx-autobuild python setup from geowebcache instructions
 Author: Jody Garnett <jody.garnett@gmail.com>
 Date: Fri Apr 28 19:07:08 2023 -0700
 4 files changed, 132 insertions(+), 14 deletions(-)
 create mode 100644 doc/en/.gitignore
 create mode 100644 doc/en/requirements.txt
Auto-merging doc/en/build.xml
Auto-merging doc/en/user/source/conf.py
CONFLICT (content): Merge conflict in doc/en/user/source/conf.py

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-2.23.x 2.23.x
# Navigate to the new working tree
cd .worktrees/backport-2.23.x
# Create a new branch
git switch --create backport-6812-to-2.23.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick cccd4a5448ce90e62f0198a5ae980df94aa4d374,ad676c6235d9b5c7b69410c8c6dc402f37af686e,e4139e5814b0425962b13952e0eb96c107fcfa2f,c732de58d35b2149293d18a39faaf14704d4dbda
# Push it to GitHub
git push --set-upstream origin backport-6812-to-2.23.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-2.23.x
```
Then, create a pull request where the `base` branch is `2.23.x` and the `compare`/`head` branch is `backport-6812-to-2.23.x`.

**Pull request merged**